File: maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.c
#include <stdio.h>
#include <assert.h>
#include "../tree_node.h"

int max(int a, int b) {
    if (a > b)
        return a;
    return b;
}

int maxDepth(struct TreeNode* root){
    if (root == NULL) {
        return 0;
    }
    return 1 + max(maxDepth(root->left), maxDepth(root->right));
}

int main() {
    struct TreeNode n0, n1, n2, n3, n4;
    init_tree_node(&n0, 3, &n1, &n2);
    init_tree_node(&n1, 9, NULL, NULL);
    init_tree_node(&n2, 20, &n3, &n4);
    init_tree_node(&n3, 15, NULL, NULL);
    init_tree_node(&n4, 7, NULL, NULL);

    int solution = 3;
    int max_depth = maxDepth(&n0);
    printf("max_depth=%d solution=%d\n", max_depth, solution);
    assert(max_depth == solution);
    printf("success!\n");

    return 0;
}

File: n150/maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree

from TreeNode import TreeNode

class Solution(object):

    def step(self, node):
        if node == None:
            return 0
        a = self.step(node.left)
        b = self.step(node.right)
        return 1 + max(a, b)
    
    def maxDepth(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        return self.step(root)
        

if __name__ == "__main__":
    sol = Solution()
    tests = [
        (
            [3,9,20,None,None,15,7],
            3,
        ),
        (
            [1,None,2],
            2
        ),
    ]
    for (nums, solution) in tests:
        tree = TreeNode.from_array(nums)
        result = sol.maxDepth(tree)
        assert result == solution, \
            f"result {result} != solution {solution}"
    print("✅ All tests passed")

